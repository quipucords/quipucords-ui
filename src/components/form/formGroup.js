import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup as PfFormGroup } from '@patternfly/react-core';
import helpers from '../../common/helpers';

const FormGroup = ({ children, error, errorMessage, helperText, id, label, ...props }) => {
  const setId = id || helpers.generateId();

  return (
    <PfFormGroup
      id={setId}
      validated={error ? 'error' : null}
      // FixMe: Applies consistent behavior to helperText regardless of it being an element or string
      helperText={helperText && <div className="pf-c-form__helper-text">{helperText}</div>}
      helperTextInvalid={
        (error && typeof errorMessage === 'string' && errorMessage) ||
        (React.isValidElement(errorMessage) && errorMessage) ||
        (errorMessage && 'Error') ||
        ''
      }
      label={label}
      {...props}
    >
      {children}
    </PfFormGroup>
  );
};

FormGroup.propTypes = {
  children: PropTypes.node.isRequired,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  errorMessage: PropTypes.oneOfType([PropTypes.node, PropTypes.bool]),
  helperText: PropTypes.node,
  id: PropTypes.string,
  label: PropTypes.node
};

FormGroup.defaultProps = {
  error: null,
  errorMessage: null,
  helperText: null,
  id: null,
  label: null
};

export { FormGroup as default, FormGroup };
