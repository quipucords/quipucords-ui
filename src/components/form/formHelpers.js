import _size from 'lodash/size';
import { helpers } from '../../common';

/**
 * Create a consistent mock event object.
 *
 * @param {object} event
 * @param {boolean} persistEvent
 * @returns {{keyCode, currentTarget, name, id: *, persist: Function, value, target}}
 */
const createMockEvent = (event, persistEvent = false) => {
  const { checked, currentTarget = {}, keyCode, persist = helpers.noop, target = {} } = { ...event };
  if (persistEvent) {
    persist();
  }

  return {
    checked,
    currentTarget,
    keyCode,
    id: currentTarget.id || currentTarget.name,
    name: currentTarget.name,
    persist,
    value: currentTarget.value,
    target
  };
};

/**
 * Confirm a string has minimum length.
 *
 * @param {string} value
 * @param {number} characters
 * @returns {boolean}
 */
const doesNotHaveMinimumCharacters = (value, characters = 1) =>
  (typeof value === 'string' && value.length < characters) || typeof value !== 'string';

/**
 * Confirm a string is not empty.
 *
 * @param {string} value
 * @returns {boolean}
 */
const isEmpty = value => (typeof value !== 'number' && _size(value) < 1) || false;

/**
 * Confirm string is the start of a path.
 *
 * @param {string} value
 * @returns {boolean}
 */
const isFilePath = value => /^\/.*$/.test(value);

/**
 * Confirm the port is valid.
 *
 * @param {string} value
 * @returns {boolean}
 */
const isPortValid = value => /^\d{1,5}$/.test(value) && value <= 65535;

const formHelpers = {
  createMockEvent,
  doesNotHaveMinimumCharacters,
  isEmpty,
  isFilePath,
  isPortValid
};

export { formHelpers as default, formHelpers, createMockEvent, doesNotHaveMinimumCharacters };
