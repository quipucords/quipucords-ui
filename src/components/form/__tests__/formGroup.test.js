import React from 'react';
import { mount } from 'enzyme';
import { FormGroup } from '../formGroup';
import { formHelpers } from '../formHelpers';

describe('FormGroup Component', () => {
  it('should render a basic component', () => {
    const props = { id: 'test' };

    const component = mount(
      <FormGroup label="test" {...props}>
        <input id="test" type="text" value="" readOnly />
      </FormGroup>
    );

    expect(component.render()).toMatchSnapshot('basic formgroup');
    expect(component.find('label').at(0).render()).toMatchSnapshot('basic label');
  });

  it('should handle multiple error message types', () => {
    const props = {
      id: 'test',
      error: true,
      errorMessage: 'lorem ipsum'
    };

    let component = mount(
      <FormGroup {...props}>
        <input id="test" type="text" value="" readOnly />
      </FormGroup>
    );

    expect(component.render()).toMatchSnapshot('string error message');

    props.errorMessage = true;

    component = mount(
      <FormGroup {...props}>
        <input id="test" type="text" value="" readOnly />
      </FormGroup>
    );

    expect(component.render()).toMatchSnapshot('boolean error message');

    props.errorMessage = <span>lorem ipsum</span>;

    component = mount(
      <FormGroup {...props}>
        <input id="test" type="text" value="" readOnly />
      </FormGroup>
    );

    expect(component.render()).toMatchSnapshot('node error message');
  });

  it('should have isEmpty validation', () => {
    expect(formHelpers.isEmpty(undefined)).toBe(true);
    expect(formHelpers.isEmpty(null)).toBe(true);
    expect(formHelpers.isEmpty('')).toBe(true);
    expect(formHelpers.isEmpty('lorem')).toBe(false);
    expect(formHelpers.isEmpty([])).toBe(true);
    expect(formHelpers.isEmpty(['lorem'])).toBe(false);
    expect(formHelpers.isEmpty({})).toBe(true);
    expect(formHelpers.isEmpty({ lorem: 'ipsum' })).toBe(false);
  });

  it('should have doesNotHaveMinimumCharacters validation', () => {
    expect(formHelpers.doesNotHaveMinimumCharacters('', 5)).toBe(true);
    expect(formHelpers.doesNotHaveMinimumCharacters('test test', 5)).toBe(false);
  });

  it('should have isPortValid validation', () => {
    expect(formHelpers.isPortValid('lorem')).toBe(false);
    expect(formHelpers.isPortValid(-1)).toBe(false);
    expect(formHelpers.isPortValid(65536)).toBe(false);
    expect(formHelpers.isPortValid('65536')).toBe(false);
    expect(formHelpers.isPortValid(65535)).toBe(true);
  });
});
