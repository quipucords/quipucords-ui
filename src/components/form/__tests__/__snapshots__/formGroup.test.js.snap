// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup Component should handle multiple error message types: boolean error message 1`] = `
<div
  class="pf-c-form__group"
  id="test"
>
  <div
    class="pf-c-form__group-control"
  >
    <input
      id="test"
      readonly=""
      type="text"
      value=""
    />
    <div
      aria-live="polite"
      class="pf-c-form__helper-text pf-m-error"
      id="undefined-helper"
    >
      Error
    </div>
  </div>
</div>
`;

exports[`FormGroup Component should handle multiple error message types: node error message 1`] = `
<div
  class="pf-c-form__group"
  id="test"
>
  <div
    class="pf-c-form__group-control"
  >
    <input
      id="test"
      readonly=""
      type="text"
      value=""
    />
    <span>
      lorem ipsum
    </span>
  </div>
</div>
`;

exports[`FormGroup Component should handle multiple error message types: string error message 1`] = `
<div
  class="pf-c-form__group"
  id="test"
>
  <div
    class="pf-c-form__group-control"
  >
    <input
      id="test"
      readonly=""
      type="text"
      value=""
    />
    <div
      aria-live="polite"
      class="pf-c-form__helper-text pf-m-error"
      id="undefined-helper"
    >
      lorem ipsum
    </div>
  </div>
</div>
`;

exports[`FormGroup Component should render a basic component: basic formgroup 1`] = `
<div
  class="pf-c-form__group"
  id="test"
>
  <div
    class="pf-c-form__group-label"
  >
    <label
      class="pf-c-form__label"
    >
      <span
        class="pf-c-form__label-text"
      >
        test
      </span>
    </label>
     
  </div>
  <div
    class="pf-c-form__group-control"
  >
    <input
      id="test"
      readonly=""
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`FormGroup Component should render a basic component: basic label 1`] = `
<label
  class="pf-c-form__label"
>
  <span
    class="pf-c-form__label-text"
  >
    test
  </span>
</label>
`;
